### ISSUE TITLE
[Feat] Discrete Gaussian kernel (gaussian) should be adapted for discrete filtering

### ISSUE THREAD

**Issue by tvercaut** [kornia/kornia#559]
## 🐛 Bug
The [current Gaussian kernel](https://github.com/kornia/kornia/blob/9a49b3c473fbbef6df1e0f245f8e99e36f6fec6a/kornia/filters/kernels.py#L19) simply samples the continuous Gaussian function:
```python
def gaussian(window_size, sigma):
    x = torch.arange(window_size).float() - window_size // 2
    if window_size % 2 == 0:
        x = x + 0.5
    gauss = torch.exp((-x.pow(2.0) / float(2 * sigma ** 2)))
    return gauss / gauss.sum()
```

## Expected behavior
Discrete Gaussian filtering should be compatible with scale-space theory and rely on modified Bessel function for its kernel weights. See e.g.:
https://en.wikipedia.org/wiki/Scale_space_implementation#The_discrete_Gaussian_kernel

## Additional context
Note that this is a similar issue to that raised in MONAI:
https://github.com/Project-MONAI/MONAI/issues/363

I came here while looking for a PyTorch implementation of a Discrete Gaussian filter but realised this one was also sampling the continuous Gaussian function.

ITK implements such a behaviour in itk::GaussianOperator:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkGaussianOperator.hxx

[Wikipedia](https://commons.wikimedia.org/wiki/File:Discrete_Gaussian_kernel.svg) also provides a basic python implementation (albeit relying on scipy and without normalisation):
```python
from scipy.special import iv
import numpy as np

def discrete_gaussian_kernel(n, t):
    T = np.exp(-t) * iv(n, t)
    return T
```

**Comment by edgarriba** [comment]
@tvercaut thanks for pointing this out. I wasn't aware of this problem. How will this affect to the actual implementation ? From what I understand while reading in wikipedia seems that methods using discrete kernel can be more robust during differentiation. If that's true, why not supporting it.

To implement this is in pytorch we need to double check that we can reproduce [`scipy.iv`](https://github.com/kornia/kornia/issues/559 ) as well. After a quick search I found some related comments on that already from pytorch community showing already some ideas. [link1](https://discuss.pytorch.org/t/modified-bessel-function-of-order-0/18609) [link2](https://github.com/nicola-decao/s-vae-pytorch/blob/671c46c0772a2c92a602252311864c48c816bc44/hyperspherical_vae/ops/ive.py). Just found that also JAX implements [Modified Bessel kernel](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.special.i1e.html#jax.scipy.special.i1e)

Besides, we were discussing with @wyli to potentially merge some functionalities under a common API. Probably. this could one first step for this initiative.

**Comment by ducha-aiki** [comment]
I will check the actual difference. In case of we are doing the change, we also would need to modify our spatial_gradient kernels. 
And it should help with local features localization precision as well.

**Comment by tvercaut** [comment]
As shown in the wikipedia image, the difference is subtle but noticeable especially for small sigmas.

Thinking a bit further, I realise some use cases might need to get the  gradient with respect to sigma. In that scenario, I assume a full pytorch implementation (not relying on scipy) is important. Also, it would be important to check the number stability of such gradient as the involved operations become more convoluted.

Maybe having both the sampled Gausian kernel and the proper Gaussian kernel makes sense.

**Comment by ducha-aiki** [comment]
Regardless of the gradients, we cannot rely on scipy anyway - our policy is to be PyTorch our only dependency.

**Comment by tvercaut** [comment]
Great, then porting a set of modified Bessel functions would be the way to go. The code in ITK for these is self contained and could thus be a good start:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkGaussianOperator.hxx

**Comment by edgarriba** [comment]
@tvercaut great ! how does compare scipy vs itk implementation ? I would target scipy behavior since seems to be a more popular package.

**Comment by wyli** [comment]
> @tvercaut great ! how does compare scipy vs itk implementation ? I would target scipy behavior since seems to be a more popular package.

`scipy.special.iv` is also interesting https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.iv.html
https://github.com/scipy/scipy/issues/1030

**Comment by edgarriba** [comment]
@wyli @tvercaut do you think you might have some time to implement this feature. We can give support for integration

**Comment by wyli** [comment]
> do you think you might have some time to implement this feature. We can give support for integration

Hi @edgarriba I'm pushing a refined gaussian1d to monai,
but this implementation is not differentiable https://github.com/Project-MONAI/MONAI/pull/1097

**Comment by ducha-aiki** [comment]
@wyli I don't see why the implementation is not differentiable. I mean, kernel generation does not to be differentiable, we just need weight to do convolution then.

**Comment by wyli** [comment]
> @wyli I don't see why the implementation is not differentiable. I mean, kernel generation does not to be differentiable, we just need weight to do convolution then.

Yes, it is differentiable :) there were some errors in the recursion, the latest commit fixes it and seems to work fine. I'm adding some unit tests about the gradient to confirm it.

**Comment by edgarriba** [comment]
@wyli @tvercaut do you already have any working real use cases showing the difference for using continuous and discrete kernels ?

**Comment by tvercaut** [comment]
@edgarriba If you only care about smoothing an image somewhat, any bell shaped kernel will do the trick. For many (most?) use cases, I assume this won't make any noticeable difference.

If you need more fine-grained scale-space analysis, the difference between the sampled Gaussian and the discrete Gaussian however can start to matter. For example applying N times a Gaussian with standard deviation σ may not be the same as aplying a Gaussian of standard deviation  √N·σ with the sampled Gaussian. It becomes also important when looking at the stability under a derivative operators for example (typically Laplacian). If you are intersted, I encourage you to read the Lindeberg paper for a good coverage: https://doi.org/10.1109/34.49051

**Comment by edgarriba** [comment]
@tvercaut I'll checkout thanks :)
@wyli we can try to generalize the functionalities and have `kornia.filter.gaussian` and `kornia.filter.gaussian_discrete` ? We are releasing in a month or v0.5. We could introduce this as part of kornia by that time in case you consider that can be useful.

**Comment by wyli** [comment]
> @tvercaut I'll checkout thanks :)
> @wyli we can try to generalize the functionalities and have `kornia.filter.gaussian` and `kornia.filter.gaussian_discrete` ? We are releasing in a month or v0.5. We could introduce this as part of kornia by that time in case you consider that can be useful.

sure, I can try to create a PR as a non-breaking new feature to kornia, or help review a PR if someone is already working on this ticket

**Comment by edgarriba** [comment]
@wyli that sounds good we could try to target to include this feature for v0.5
